optaplanner-wb-screens/optaplanner-wb-guided-rule-editor/optaplanner-wb-guided-rule-editor-client/src/main/java/org/optaplanner/workbench/screens/guidedrule/client/widget/ConstraintMatchInputWidgetBlurHandler.java

```diff
@@ -18,20 +18,75 @@
 
 import com.google.gwt.event.dom.client.BlurEvent;
 import com.google.gwt.event.dom.client.BlurHandler;
+import org.jboss.errai.ui.client.local.spi.TranslationService;
+
+import static org.optaplanner.workbench.screens.guidedrule.client.resources.i18n.GuidedRuleEditorConstants.ConstraintMatchInputWidgetBlurHandler_DoubleValueParsingError;
+import static org.optaplanner.workbench.screens.guidedrule.client.resources.i18n.GuidedRuleEditorConstants.ConstraintMatchInputWidgetBlurHandler_EmptyValuesAreNotAllowedForModifyScore;
+import static org.optaplanner.workbench.screens.guidedrule.client.resources.i18n.GuidedRuleEditorConstants.ConstraintMatchInputWidgetBlurHandler_IntegerValueParsingError;
+import static org.optaplanner.workbench.screens.guidedrule.client.resources.i18n.GuidedRuleEditorConstants.ConstraintMatchInputWidgetBlurHandler_LongValueParsingError;
 
 public class ConstraintMatchInputWidgetBlurHandler implements BlurHandler {
 
     private ConstraintMatchInputWidget widget;
+    private TranslationService translationService;
+    private String scoreHolderType;
 
-    public ConstraintMatchInputWidgetBlurHandler(ConstraintMatchInputWidget widget) {
+    public ConstraintMatchInputWidgetBlurHandler(final ConstraintMatchInputWidget widget,
+                                                 final TranslationService translationService,
+                                                 final String scoreHolderType) {
         this.widget = widget;
+        this.translationService = translationService;
+        this.scoreHolderType = scoreHolderType;
     }
 
     @Override
     public void onBlur(BlurEvent event) {
-        if (widget.getConstraintMatchValue() == null || widget.getConstraintMatchValue().trim().isEmpty()) {
-            widget.showEmptyValuesNotAllowedError();
+        String inputValue = widget.getConstraintMatchValue();
+        if (inputValue == null || inputValue.trim().isEmpty()) {
+            widget.showError(translationService.getTranslation(ConstraintMatchInputWidgetBlurHandler_EmptyValuesAreNotAllowedForModifyScore));
         } else {
+            inputValue = inputValue.trim();
+            if (inputValue.matches("-?\\s*\\d+(\\.\\d+)?.*")) {
```

Comment:
@manstis Well, BlurHandler is not ideal indeed. I tried to change the code locally to get instant validation using KeyUpHandler and it feels better. Is that what you had on your mind? Saving invalid input is possible, you've been warned.